### PR TITLE
Invoke component cleanup in the Operator controller on Submariner deletion

### LINKED ISSUE #1722: Invoke component cleanup in the Operator controller on Submariner deletion
Add a finalizer to the Submariner resource to initiate the dataplane cleanup by creating `DaemonSets` for the gateway engine, route agent et al to run a [one-time task](http://blog.itaysk.com/2017/12/26/the-single-use-daemonset-pattern-and-prepulling-images-in-kubernetes) using the same image but passing an UNINSTALL environment variable.

Task for https://github.com/submariner-io/enhancements/issues/50

### PR DESCRIPTION
**Pull request by tpantelis**
Create `DaemonSets`/`Deployments` to uninstall the **submariner-gateway**, **submariner-routeagent**, **submariner-globalnet**, and **submariner-networkplugin-syncer** components as outlined in #1722. See individual commits for details.

Fixes #1722

### CODE REVIEW

File: controllers/submariner/submariner_controller_test.go [hunk 1]
```diff
@@ -20,78 +20,23 @@ package submariner_test
...
-	"k8s.io/client-go/kubernetes/scheme"
-	controllerClient "sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/fake"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 const (
-	submarinerName          = "submariner-gateway"
+	submarinerName          = "submariner"
 	submarinerNamespace     = "submariner-operator"
```
Comment (Jaanki) [inline controllers/submariner/submariner_controller_test.go]: These 2 constants are defined in `internal/constants/constants.go` as well. That could be used here.

File: controllers/submariner/cleanup.go [hunk 2]
```diff
@@ -39,14 +39,21 @@ import (
...
 	// First, delete the regular DaemonSets/Deployments and ensure all their pods are cleaned up.
-	objsToDelete := []client.Object{newDaemonSet(names.GatewayComponent, instance.Namespace)}
 	for _, obj := range objsToDelete {
 		err := r.ensureDeleted(ctx, obj)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
+	}
 
+	// Next, delete ensure all their pods are cleaned up.
```
Comment (Jaanki) [inline controllers/submariner/cleanup.go]: // Next, delete `and` ensure all their pods are cleaned up.?

File: controllers/submariner/submariner_controller_test.go [hunk 3]
```diff
@@ -35,9 +35,8 @@ import (
 )
 
 const (
-	submarinerName          = "submariner"
-	submarinerNamespace     = "submariner-operator"
-	routeAgentDaemonSetName = "submariner-routeagent"
+	submarinerName      = "submariner"
```
Comment (Jaanki) [inline controllers/submariner/submariner_controller_test.go]: reuse these from `internal/constants/constants.go`?

**Review by Jaanki — CHANGES_REQUESTED** [review]

File: controllers/submariner/submariner_controller_test.go (hunk 1 above)
Comment (tpantelis) [reply controllers/submariner/submariner_controller_test.go]: These predated this PR. But the controller code does not/should not rely on the global constants (the namespace at least) so neither should the test code. In fact we should use values that differ from the global constants to ensure the controller code does not use them. Changed with https://github.com/submariner-io/submariner-operator/pull/1803/commits/06112225d96942e0e0a214c4d494970209650b91.

File: controllers/submariner/cleanup.go (hunk 2 above)
Comment (tpantelis) [reply controllers/submariner/cleanup.go]: yeah that's fixed in a subsequent commit.

File: controllers/submariner/submariner_controller_test.go (hunk 3 above)
Comment (tpantelis) [reply controllers/submariner/submariner_controller_test.go]: Same as above.

File: controllers/submariner/submariner_controller_test.go [hunk 4]
```diff
@@ -239,14 +238,16 @@ func testDeletion() {
 
 		t.initClientObjs = append(t.initClientObjs,
 			t.newDaemonSet(names.GatewayComponent),
-			t.newPodWithLabel("app", names.GatewayComponent))
+			t.newPodWithLabel("app", names.GatewayComponent),
+			t.newDaemonSet(names.RouteAgentComponent))
 	})
 
-	It("should uninstall the component resources and remove the finalizer", func() {
+	FIt("should uninstall the component resources and remove the finalizer", func() {
```
Comment (dfarrell07) [inline controllers/submariner/submariner_controller_test.go]: Typo? s/FIt/It

Comment (tpantelis) [reply controllers/submariner/submariner_controller_test.go]: F = focus - I removed the F in the next commit.

Comment (dfarrell07) [reply controllers/submariner/submariner_controller_test.go]: Ah okay, thanks/learning :nerd_face: :+1:

**Review by dfarrell07 — APPROVED** [review]

File: controllers/submariner/submariner_controller_test.go (hunk 4 above)
Comment (dfarrell07) [reply controllers/submariner/submariner_controller_test.go]: (I noticed it went away btw, that's part of why I thought it was a typo)

**Review by Jaanki — APPROVED** [review]

File: controllers/submariner/np_syncer_resources.go [hunk 5]
```diff
@@ -123,3 +123,7 @@ func newNetworkPluginSyncerDeployment(cr *v1alpha1.Submariner, clusterNetwork *n
 
 	return networkPluginSyncerDeployment
 }
+
+func canInstallNetworkPluginSyncer(submariner *v1alpha1.Submariner) bool {
```
Comment (skitt) [inline controllers/submariner/np_syncer_resources.go]: Bit of a nitpick, but wouldn’t this be better as `needsNetworkPluginSyncer`?

File: controllers/submariner/submariner_controller.go [hunk 6]
```diff
@@ -239,6 +246,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 			// TODO skitt determine what we should really be doing for concurrent updates to the Submariner CR
 			// Updates fail here because the instance is updated between the .Update() at the start of the function
 			// and the status update here
+			fmt.Printf("****UPDATE: %#v", err)
```
Comment (skitt) [inline controllers/submariner/submariner_controller.go]: Is this a leftover?

**Review by skitt — CHANGES_REQUESTED** [review]

File: controllers/submariner/np_syncer_resources.go (hunk 5 above)
Comment (tpantelis) [reply controllers/submariner/np_syncer_resources.go]: Sure

File: controllers/submariner/submariner_controller.go (hunk 6 above)
Comment (tpantelis) [reply controllers/submariner/submariner_controller.go]: Crap - yes.

File: controllers/submariner/submariner_controller.go [hunk 7]
```diff
@@ -59,6 +61,7 @@ import (
 const (
 	gatewayMetricsServerPort   = 8080
 	globalnetMetricsServerPort = 8081
+	SubmarinerFinalizer        = "submariner-controller-finalizer"
```
Comment (sridhargaddam) [inline controllers/submariner/submariner_controller.go]: nit: Usually, I have seen people using finalizers like 
```
    finalizers:                                                                                                                                                                              
    - managedcluster-import-controller.open-cluster-management.io/cleanup                                                                                                                    
    - managedclusterinfo.finalizers.open-cluster-management.io                                                                                                                               
    - cluster.open-cluster-management.io/api-resource-cleanup                                                                                                                                
    - agent.open-cluster-management.io/klusterletaddonconfig-cleanup                                                                                                                         
    - open-cluster-management.io/managedclusterrole                                                                                                                                          
    - managedcluster-import-controller.open-cluster-management.io/manifestwork-cleanup                                                                                                       
    - cluster.open-cluster-management.io/submariner-agent-cleanup
```

So shall we follow the same pattern and name it something like `submariner.io/cleanup" ?

File: controllers/submariner/cleanup.go [hunk 8]
```diff
@@ -20,15 +20,148 @@ package submariner
...
+	if daemonSet.Status.ObservedGeneration == 0 || daemonSet.Status.ObservedGeneration < daemonSet.Generation {
+		log.Info("DaemonSet generation not yet observed - requeueing:", "name", daemonSet.Name,
+			"namespace", daemonSet.Namespace, "Generation", daemonSet.Generation, "ObservedGeneration", daemonSet.Status.ObservedGeneration)
+		return true, nil
+	}
+
+	if daemonSet.Status.DesiredNumberScheduled == 0 {
+		log.Info("DaemonSet has no available nodes:", "name", daemonSet.Name, "namespace", daemonSet.Namespace)
+	} else if daemonSet.Status.DesiredNumberScheduled != daemonSet.Status.NumberReady {
+		log.Info("DaemonSet not ready yet - requeueing:", "name", daemonSet.Name, "namespace", daemonSet.Namespace,
```
Comment (sridhargaddam) [inline controllers/submariner/cleanup.go]: What would happen if the admin marks one of the clusterNodes as unschedulable (ie.., kubectl cordon node-name)? Will we continue to retry as the desiredNumber will not match numberReady?

File: controllers/submariner/cleanup.go [hunk 9]
```diff
@@ -20,15 +20,148 @@ package submariner
...
+
+	return false, r.ensureDeleted(ctx, daemonSet)
+}
+
+func convertPodSpecContainersToUninstall(podSpec *corev1.PodSpec) {
+	// We're going to use the PodSpec run a one-time task by using an init container to run the task.
+	// See http://blog.itaysk.com/2017/12/26/the-single-use-daemonset-pattern-and-prepulling-images-in-kubernetes
+	// for more details.
+	podSpec.InitContainers = podSpec.Containers
+	podSpec.InitContainers[0].Env = append(podSpec.InitContainers[0].Env, corev1.EnvVar{Name: "UNINSTALL", Value: "true"})
```
Comment (sridhargaddam) [inline controllers/submariner/cleanup.go:328]: So is the init container responsible to do the necessary cleanup on the respective node?
Where are we passing the script (or the code) to perform the cleanup?

**Review by sridhargaddam — COMMENTED** [review]
Thanks for the PR @tpantelis. Its a jumbo PR but the individual commits helped.

This PR/feature deserves a tech-exchange session ;-)

File: controllers/submariner/cleanup.go (hunk 8 above)
Comment (tpantelis) [reply controllers/submariner/cleanup.go]: I don't know how K8s handles an unschedulable node, ie is it counted in `DesiredNumberScheduled` (logically I wouldn't think so)? Either way we will eventually timeout and delete the `DaemonSet`.

File: controllers/submariner/cleanup.go (hunk 9 above)
Comment (tpantelis) [reply controllers/submariner/cleanup.go:328]: > So is the init container responsible to do the necessary cleanup on the respective node? 

Yes

> Where are we passing the script (or the code) to perform the cleanup?

I copy the normal container to the init container (L149) so it's the same image and params except I add the "UNINSTALL" env var.

File: controllers/submariner/submariner_controller.go (hunk 7 above)
Comment (tpantelis) [reply controllers/submariner/submariner_controller.go]: Done with https://github.com/submariner-io/submariner-operator/pull/1803/commits/31cfc9d3e374c0eae00bd6c06f5f2799668957f5

File: controllers/submariner/cleanup.go (hunk 8 above)
Comment (tpantelis) [reply controllers/submariner/cleanup.go]: According to [this](https://kubernetes.io/docs/concepts/architecture/nodes/), "_Marking a node as unschedulable prevents the scheduler from placing new pods onto that Node_" so i'd think `DesiredNumberScheduled` wouldn't reflect it.

**Review by sridhargaddam — APPROVED** [review]
Thanks @tpantelis

**Review by skitt — APPROVED** [review]

**Pull request merged**